nessie/run.go

```diff
@@ -0,0 +1,124 @@
+package nessie
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/thanhpk/randstr"
+	"github.com/treeverse/lakefs/api"
+	"github.com/treeverse/lakefs/api/gen/client"
+	"github.com/treeverse/lakefs/api/gen/client/setup"
+	"github.com/treeverse/lakefs/api/gen/models"
+	"github.com/treeverse/lakefs/logging"
+)
+
+// Config contains configuration needed for running the tests
+type Config struct {
+	// BaseURL is the base address of the lakeFS endpoint
+	BaseURL string
+
+	// BucketPath is the full path to the s3 bucket
+	BucketPath string
+}
+
+// Run runs system tests and reports on failures
+func Run(ctx context.Context, config Config) error {
+	// initialize the env/repo
+	logger := logging.FromContext(ctx)
+	bucketPath := fmt.Sprintf("%s/%s", config.BucketPath, time.Now().UTC().Format("2006/01/02/150405"))
+	logger.WithField("bucketPath", bucketPath).Info("Starting nessie run")
+
+	cl := client.NewHTTPClientWithConfig(strfmt.Default, &client.TransportConfig{
+		Host:     config.BaseURL,
+		BasePath: client.DefaultBasePath,
+		Schemes:  []string{"http"},
+	})
+
+	expBackoff := backoff.NewExponentialBackOff()
+	if err := backoff.Retry(func() error {
+		_, err := cl.HealthCheck.HealthCheck(nil)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, expBackoff); err != nil {
+		return fmt.Errorf("api endpoint isn't ready: %w", err)
+	}
+
+	adminUserName := "nessie"
+	res, err := cl.Setup.SetupLakeFS(&setup.SetupLakeFSParams{
+		User: &models.Setup{
+			DisplayName: &adminUserName,
+		},
+		Context: ctx,
+	})
+	if err != nil {
+		return fmt.Errorf("lakefs setup request failed: %w", err)
+	}
+
+	logger.WithField("accessKeyID", res.Payload.AccessKeyID).
+		WithField("accessSecretKey", res.Payload.AccessSecretKey).
+		Info("Cluster setup successfully")
+
+	c, err := api.NewClient(fmt.Sprintf("http://%s/%s", config.BaseURL, client.DefaultBasePath), res.Payload.AccessKeyID, res.Payload.AccessSecretKey)
+	if err != nil {
+		return fmt.Errorf("failed to setup client: %w", err)
+	}
+
+	if err := c.CreateRepository(ctx, &models.RepositoryCreation{
```

Comment:
Kind of a big function, dealing with 2 separate concerns - setting up an environment to work with, and starting from this line, the actual test itself.

I suggest modeling it using Go's test framework - you get reporting, coverage, etc out of the box.
A running system is private case for a fixture.

With `go test` you still get the benefits you wanted but it requires no special infra to setup, just a `go test -tags system`.
you'll also get the benefit of `go test -race` which should deterministically fail many cases where you'd otherwise get flakiness, so easier to reproduce as well.